### PR TITLE
[Core] NurbsSurface knot refinement

### PR DESCRIPTION
**Pull request by tteschemacher**
**Description**
Adds a refinement utility and according tests.

### CODE REVIEW

File: kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        const SizeType nb_cp_v = rGeometry.PointsNumberInDirection(1);
+
+        const Vector& knots_u_old = rGeometry.KnotsU();
+
+        const Vector& weights_old = rGeometry.Weights();
+
+        const SizeType nb_knots_u_old = knots_u_old.size();
+
+        const SizeType a = NurbsUtilities::GetUpperSpan(degree_u, knots_u_old, *knots_u_old.begin());
+        const SizeType b = NurbsUtilities::GetUpperSpan(degree_u, knots_u_old, *knots_u_old.end());
```
Comment (oberbichler) [inline kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp]: AFAIK `.end()` is pointing to `vector[n + 1]`. Maybe use `front` and `back` ([see](https://github.com/oberbichler/ANurbs/blob/e508daa256a292850e25ca48bc74ee8cfbe4efd2/include/anurbs/Algorithm/KnotRefinement.h#L37-L38)).

File: kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,329 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
```
Comment (oberbichler) [inline kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp:9]:
```suggestion
//  Ported from the ANurbs library (https://github.com/oberbichler/ANurbs)
//
```

File: kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp (hunk 1 above)
Comment (tteschemacher) [reply kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp]: thanks for that hint. Indeed `.end()` should not be correct. The UBLAS vector however does not has `.back()`

File: kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp (hunk 2 above)
Comment (tteschemacher) [reply kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp:9]: I added references to the respective comments.

**Review by manuelmessmer — APPROVED** [review]

**Comment by tteschemacher** [comment]
Awesome. Thanks all for checking!

**Pull request merged**